phpBB/phpbb/db/result_iterator.php

```diff
@@ -0,0 +1,254 @@
+	 */
+	public function __construct(ResultStatement $statement)
+	{
+		$this->data = $statement;
+		$this->id = self::$s_id++;
+	}
+
+	/**
+	 * Iterator destructor.
+	 */
+	public function __destruct()
+	{
+		if ($this->data instanceof ResultStatement)
+		{
+			$this->data->closeCursor();
+		}
+	}
+
+	/**
+	 * Moves the iterator to the specified position.
+	 *
+	 * @param int $position The position to which the iterator should be moved to.
+	 */
+	public function seek($position)
+	{
+		$this->target_position = $position;
+	}
+
+	/**
+	 * Rewinds the iterator to the starting position.
+	 *
+	 * @return void
+	 */
+	public function rewind()
+	{
+		$this->seek(0);
+	}
+
+	/**
+	 * Returns the current value to which the iterator points.
+	 *
+	 * Note: If the current state of the iterator was not validated by calling `valid()` the return value is
+	 * 		 undefined.
+	 *
+	 * @return array|null
+	 */
+	public function current()
+	{
+		return $this->current;
+	}
+
+	/**
+	 * Returns the current iterator position.
+	 *
+	 * Note: If the current state of the iterator was not validated by calling `valid()` the return value is
+	 * 		 undefined.
+	 *
+	 * @return int The current iterator position.
+	 */
+	public function key()
+	{
+		return $this->position;
+	}
+
+	/**
+	 * Advances the iterator to the next element.
+	 *
+	 * @return void
+	 */
+	public function next()
+	{
+		++$this->target_position;
+	}
+
+	/**
+	 * Returns whether or not the state of the iterator is valid.
+	 *
+	 * Note: Determining whether the current position is valid requires actually fetching the data.
+	 *
+	 * @return bool Returns whether the current positions is valid.
+	 */
+	public function valid()
+	{
+		$this->iterate();
+		return is_array($this->current);
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function fetch_all()
+	{
+		if (!($this->data instanceof ResultStatement))
+		{
+			return $this->data;
+		}
+
+		$rows = $this->data->fetchAll(FetchMode::ASSOCIATIVE);
+		$this->data->closeCursor();
+		$this->data = $rows;
+
+		return $this->data;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function invalidate()
+	{
+		if ($this->data instanceof ResultStatement)
+		{
+			$this->data->closeCursor();
+		}
+
+		$this->data = [];
+		$this->current = null;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function get_id()
+	{
+		return $this->id;
+	}
+
+	/**
+	 * Move the iterator and fetch the current element.
+	 *
+	 * @return void
+	 */
+	private function iterate()
+	{
+		switch ($this->position <=> $this->target_position)
+		{
+			case 1: // $this->position > $this->target_position
+				$this->iterate_backward();
+			break;
+			case -1: // $this->position < $this->target_position
+				$this->iterate_forward();
+			break;
+		}
+
+		$this->position = $this->target_position;
+	}
+
+	/**
+	 * Fetch an element before the current iterator position.
+	 *
+	 * @return void
+	 */
+	private function iterate_backward()
```

Comment:
As of now, this method is just horse shit mixed with snake oil, as `fetch_all()` doesn't fetch all but all remaining rows from the current cursor position. The problem here is saving all rows (which is one option to actually support going backwards) would add some overhead on all queries. I'm not sure if breaking BC here would make more sense, as I guess this might be some functionality that is not used at all by anyone (certainly not by the core).

@marc1706 @Derky @Nicofuma thoughts?

Reply:
IMHO if there is no good way to achieve this then don't. Especially if it's not even used in the core. We should however make sure to properly document this BC break, e.g. in the documentation. I can prepare a page for this (similar to how symfony is documenting those changes: https://github.com/symfony/symfony/blob/master/UPGRADE-5.0.md).

Reply:
The only solution I see is to store all fetched rows in the iterator. I doubt that there would be any other solution to this.

Reply:
Although we could reexecute the query as well when backward iteration is called for, thus only having any overhead when it is absolutely required.

Reply:
Is it not possible to save queries that are made per page in a protected property? This will also allow extensions to access SQL queries that are already made by the core, without having to rerun them if they're not cached.
For example saving the SQL queries in `$queries` and all the SQL results in `$results`.
```php
protected $queries = [
    0 => 'SELECT * FROM `phpbb_forums`,
];

protected $results = [
    0 => [
        0 => ['forum_id => 1', '...' => '...'],
    ],
];
```
So you can iterate over them and check before querying to see if it already exists.

Reply:
I don't think we would want to add any new features to this, as the goal here is to keep BC as much as possible. Also, if you re-execute a query, chances are it comes from the cache which should be much faster. Thirdly, extensions should not really execute the same queries as that data probably is accessible to them from somewhere.

Reply:
Well, that's not completely true as a lot of queries in the core are not cached and not accessible for extensions, if not in tedious ways. But it was just an idea.

Reply:
Extension authors can add events if they want to.